CASC/PortfolioMode.cpp

```diff
@@ -353,72 +347,91 @@ void PortfolioMode::getSchedules(Property& prop, Schedule& quick, Schedule& fall
   }
 }
 
+bool PortfolioMode::runSchedule(Shell::Property *property, Schedule schedule) {
+  CALL("PortfolioMode::runSchedule");
 
-// Simple one-after-the-other priority.
-float PortfolioProcessPriorityPolicy::staticPriority(vstring sliceCode)
-{
-  static float priority = 0.;
-  priority += 1.;
-  return priority;
-}
-
-//should never be called
-float PortfolioProcessPriorityPolicy::dynamicPriority(pid_t pid)
-{
-  ASSERTION_VIOLATION;
-  return 0.;
-}
-
-PortfolioSliceExecutor::PortfolioSliceExecutor(PortfolioMode *mode)
-  : _mode(mode)
-{}
-
-void PortfolioSliceExecutor::runSlice(vstring sliceCode,int remainingTime)
-{
-  CALL("PortfolioSliceExecutor::runSlice");
+  Schedule::BottomFirstIterator it(schedule);
+  Set<pid_t> processes;
+  bool success = false;
+  int remainingTime;
+  while(Timer::syncClock(), remainingTime = env.remainingTime() / 100, remainingTime > 0)
+  {
+    // running under capacity, wake up more tasks
+    while(processes.size() < _numWorkers)
+    {
+      // after exhaustion we replace the schedule
+      // by copies with x2 time limits and do this forever
+      if(!it.hasNext()) {
+        Schedule next;
+        getExtraSchedules(*property, schedule, next, false, 2);
+        schedule = next;
+        it = Schedule::BottomFirstIterator(schedule);
+      }
+      ALWAYS(it.hasNext());
+
+      vstring code = it.next();
+      pid_t process = Multiprocessing::instance()->fork();
+      ASS_NEQ(process, -1);
+      if(process == 0)
+      {
+        runSlice(code, remainingTime);
+        ASSERTION_VIOLATION; // should not return
+      }
+      ALWAYS(processes.insert(process));
+    }
 
-  vstring chopped;
-  int sliceTime = _mode->getSliceTime(sliceCode, chopped);
+    bool exited, signalled;
```

Comment:
The logic is simplified a bit from the original since we can no longer have "stopped" processes.